graphql_compiler/compiler/emit_sql.py

```diff
@@ -239,19 +340,152 @@ def start_global_operations(self):
+        #   array_agg(InnerVertex.fold_output_column) AS fold_output
+        # FROM OuterVertex
+        # INNER JOIN ... <- INNER JOINs compiled during Unfold block
+        # ON ...
+        #          ...
+        # INNER JOIN InnerVertex <- INNER JOINs compiled during Unfold block
+        # ON ...
+        # GROUP BY OuterVertex.uuid
+
+        if self._in_fold:
+            raise AssertionError(u'Reached an invalid fold state.')
+
+        # begin the fold
+        self._in_fold = True
+        # outer vertex is whatever vertex occurs immediately before the fold
+        self._fold_outer_vertex_location = self._current_location
+        self._fold_inner_vertex_location = fold_scope_location
+
+        # basic info about the outer vertex and inner vertex
+        inner_vertex_table = self._sql_schema_info.vertex_name_to_table[
+            self._ir.query_metadata_table.get_location_info(fold_scope_location).type.name
+        ].alias()
+        outer_vertex_table = self._current_alias.alias()
+        inner_vertex_type = self._ir.query_metadata_table \
+            .get_location_info(fold_scope_location) \
+            .type \
+            .name
+        outer_vertex_type = self._current_classname
+
+        self._aliases[
+            (self._current_location.query_path, self._fold_inner_vertex_location)
+        ] = inner_vertex_table
+
+        # collect edge information to join the fold subquery to the main selectable
+        edge_direction = fold_scope_location.fold_path[0][0]
+        edge_name = fold_scope_location.fold_path[0][1]
+        full_edge_name = '{}_{}'.format(edge_direction, edge_name)
+
+        edge = self._sql_schema_info.join_descriptors[
+            self._current_classname][full_edge_name]
+
+        # we will use this to make sure that the join inside
+        # the subquery uses one alias of the current table
+        # and that the join outside the subquery uses another alias
+        join_info = (edge,
+                     outer_vertex_table,
+                     inner_vertex_type,
+                     self._current_alias)
+
+        self._fold = SQLFoldObject(inner_vertex_table,
+                                   outer_vertex_table,
+                                   inner_vertex_type,
+                                   outer_vertex_type,
+                                   join_info)
+        # find outputs for this fold in self._all_folded_outputs and add to self._fold_outputs
+        if fold_scope_location.fold_path in self._all_folded_outputs:
+            for fold_output in self._all_folded_outputs[fold_scope_location.fold_path]:
+                if fold_output.field == '_x_count':
+                    raise NotImplementedError(u'_x_count not implemented in SQL')
+
+                intermediate_fold_output_name = self._alias_generator.gen_output_column()
+                self._fold_output_column_aliases[fold_output] = intermediate_fold_output_name
```

Comment:
This is suspicious for two reasons (maybe I don't understand it):
1. I'm guessing we generate `intermeiate_fold_output_name` because `fold_output`s can collide from different outputs on different scopes inside the fold, but on fields with the same name. If that's the case, can you try to find an example query where that happens? I'm pretty sure the rules and constraints on `@fold` in the README rule out that case.
2. What's the point of generating unique names if we're using `fold_output` as a unique key in a dictionary? If `fold_output` is not unique, we will get collisions in this dictionary. Furthermore this `self._fold_output_column_aliases` is shared between all folds we process, and we can get into some really nasty bugs if information about one fold overrides information about another fold in the query.

Reply:
The primary motivation is that if you don't provide some name then the resulting output column:

```
    sqlalchemy.func.array_agg(
         self._fold.inner_vertex_alias.c[fold_output.field]
     )
```

the resulting column is given an `anon_label` instead of a `label`. The `anon_label` isn't actually generated until the column is selected (to my knowledge). This would make it difficult to associate a fold scope location with stable alias - but we need to know the stable alias because at line 740 in expressions.to_sql() we index directly into the fold subquery table by the name of the column being selected. I don't really see a way to get this to work without explicitly providing _some_ label. That said, providing globally unique names, whatever those names are makes sure that we can always index into the table containing this column and get precisely that column back.

I don't know if collisions is actually an issue because fold_output should be unique to a specific fold and a specific output column? I'm not sure though, could you elaborate on that reasoning?

Reply:
Great, `fold_output` is unique. Can we make `intermediate_fold_output_name = fold_output`? I think that addresses the `anon_label` issue because we explicitly name our columns. It also removes the need for `self._fold_output_column_aliases` entirely. Let me know if I'm missing something.

Reply:
fold_output looks like, e.g. `FoldScopeLocation(Location(('Animal',), None, 1), (('out', 'Animal_ParentOf'),), field=name)` I think I could write `intermediate_fold_output_name = fold_output.field` without getting any errors as long as no table in the schema has duplicate fields and as long as the same field isn't output twice in the fold. Those two cases would be validation errors right?

Reply:
Yes, the `compiler_frontend` should complain if any of those two cases happens.